src/dataqueue/queue.cc

```diff
@@ -0,0 +1,1221 @@
+  //
+  // If the entry is idempotent, a size should always be available.
+  Maybe<uint64_t> size() const override { return data_queue_->size(); }
+
+  // When true, multiple reads on the object must produce the exact
+  // same data or the reads will fail. Some sources of entry data,
+  // such as streams, may not be capable of preserving idempotency
+  // and therefore must not claim to be. If an entry claims to be
+  // idempotent and cannot preserve that quality, subsequent reads
+  // must fail with an error when a variance is detected.
+  bool isIdempotent() const override { return data_queue_->isIdempotent(); }
+
+  void MemoryInfo(node::MemoryTracker* tracker) const override {
+    tracker->TrackField("data_queue", data_queue_);
+  }
+
+  DataQueue& getDataQueue() { return *data_queue_; }
+
+  SET_MEMORY_INFO_NAME(DataQueueEntry);
+  SET_SELF_SIZE(DataQueueEntry);
+
+ private:
+  std::shared_ptr<DataQueue> data_queue_;
+
+  class ReaderImpl : public DataQueue::Reader {
+   public:
+    explicit ReaderImpl(std::shared_ptr<DataQueue::Reader> inner)
+        : inner_(std::move(inner)) {}
+
+    int Pull(DataQueue::Reader::Next next,
+             int options,
+             DataQueue::Vec* data,
+             size_t count,
+             size_t max_count_hint) override {
+      return inner_->Pull(
+          std::move(next), options, data, count, max_count_hint);
+    }
+
+    SET_NO_MEMORY_INFO()
+    SET_MEMORY_INFO_NAME(ReaderImpl)
+    SET_SELF_SIZE(ReaderImpl)
+
+   private:
+    std::shared_ptr<DataQueue::Reader> inner_;
+  };
+};
+
+// ============================================================================
+
+// An FdEntry reads from a file descriptor. A check is made before each read
+// to determine if the fd has changed on disc. This is a best-effort check
+// that only looks at file size, creation, and modification times. The stat
+// check is also async, so there's a natural race condition there where the
+// file could be modified between the stat and actual read calls. That's
+// a tolerable risk here. While FdEntry is considered idempotent, this race
+// means that it is indeed possible for multiple reads to return different
+// results if the file just happens to get modified.
+class FdEntry final : public EntryBase {
+  // TODO(@jasnell, @flakey5):
+  // * This should only allow reading from regular files. No directories, no
+  // pipes, etc.
+  // * The reader should support accepting the buffer(s) from the pull, if any.
+  // It should
+  //   only allocate a managed buffer if the pull doesn't provide any.
+  // * We might want to consider making the stat on each read sync to eliminate
+  // the race
+  //   condition described in the comment above.
+ public:
+  FdEntry(Environment* env,
+          int fd,
+          uint64_t start,
+          v8::Maybe<uint64_t> end,
+          BaseObjectPtr<fs::FileHandle> maybe_file_handle =
+              BaseObjectPtr<fs::FileHandle>())
```

Comment:
The ownership semantics for `fd` should be described somewhere if there’s no `fs::FileHandle` associated with this. (There should probably always be one?)